radian: init at 0.5.11 and provide radianWrapper

jdreaver opened the pull request:
Also:
lineedit: init at 0.1.6
rchitect: init at 0.3.32

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR adds the [`radian` R console](https://pypi.org/project/radian/) along with a couple of its dependencies. It provides `radianWrapper`, similar to `rWrapper` and `rstudioWrapper`, to provide packages to be used with the `radian` shell.

It can be used like so:

```nix
radianWrapper.override {
  packages = with rPackages; [ purrr ]; 
} 
```

Here is an example of it getting built:

```sh
$ nix repl
Welcome to Nix version 2.4pre20210802_47e96bb. Type :? for help.

nix-repl> flake = builtins.getFlake (toString ./.)                                                                    

nix-repl> :b with flake.legacyPackages.x86_64-linux; radianWrapper.override { packages = with rPackages; [ purrr ]; } 
warning: error: plugin-files set after plugins were loaded, you may need to move the flag before the subcommand

this derivation produced the following outputs:
  out -> /nix/store/8hakcfw1zm7v0x5d428mk0nj67qaqprb-radian-wrapper
```

And here is the previous build getting used:

```R
$ /nix/store/8hakcfw1zm7v0x5d428mk0nj67qaqprb-radian-wrapper/bin/radian
R version 4.1.1 (2021-08-10) -- "Kick Things"
Platform: x86_64-pc-linux-gnu (64-bit)

r$> library(purrr)                                                                                                                    
r$> mtcars %>% 
          split(.$cyl) %>% # from base R 
          map(~ lm(mpg ~ wt, data = .)) %>% 
          map(summary) %>% 
          map_dbl("r.squared")          
        4         6         8                                                                                                       
0.5086326 0.4645102 0.4229655                                                                                                                               
```

In particular, the key thing here is making sure `radian` knows about `R_HOME` and `R_LIBS_SITE`, which I verified:

```R
r$> Sys.getenv("R_HOME")
[1] "/nix/store/w57r9xnsfr5snp9pdyqy7xqxq6lgbhay-R-4.1.1/lib/R"

r$> Sys.getenv("R_LIBS_SITE")
[1] "/nix/store/zvbxvifrnqwlwy9zhclbs06x08086nfb-r-boot-1.3-28/library:/nix/store/mqsrh4ykivjh49wmmba8d0gnfawlk0qi-r-class-7.3-19/library:/nix/store/m54bbpls1yj6444vyrfh9if3cacssjcl-r-MASS-7.3-54/library:/nix/store/5gxxm42zjznj2z3ymwqjyf14sblmzdrz-r-cluster-2.1.2/library:/nix/store/mh3iilyaxqq36cxfa0i6qlr761ylc3fh-r-codetools-0.2-18/library:/nix/store/j65lddbycwx009kd595924df27ly4rbp-r-foreign-0.8-81/library:/nix/store/xkdxwjaf0ryfqabxqlpfna3bigw0wb4d-r-KernSmooth-2.23-20/library:/nix/store/icw2qral5naskxl5s152dpbnic945w0c-r-lattice-0.20-44/library:/nix/store/fvk8ripxpbr6sixy6386f1aq27z43kqg-r-Matrix-1.3-3/library:/nix/store/wsnjrc3j9p4qnd79d6wdfrw8sw9nibi1-r-mgcv-1.8-35/library:/nix/store/2jsny9m90a5hi5pz56clkra903wn2cbj-r-nlme-3.1-152/library:/nix/store/a167w7vbc93iyzd4ajkg8inqj78cn61r-r-nnet-7.3-16/library:/nix/store/r0kid1yzdhx9maa3s195bx3rdhfmgrx0-r-rpart-4.1-15/library:/nix/store/nrrkjy221cgpzkqv6zlvx27pdsp49y2j-r-spatial-7.3-14/library:/nix/store/7vnys4dnpyhf3w234l23n4fpdjlv49x3-r-survival-3.2-11/library:/nix/store/5h4x9h0d5h5a6vr6ff7kgrzk16k0g71r-r-purrr-0.3.4/library:/nix/store/gc5bjx2akadw8wap86n561l4pf5vi8dx-r-magrittr-2.0.1/library:/nix/store/89ky56ijslgz3n0gr7a35vlxm14ljzrx-r-rlang-0.4.11/library"

```

I also tested that all 3 of the Python packages built properly with both Python 2 and 3:

```sh
$ nix build .#pythonPackages.lineedit 
$ nix build .#python3Packages.lineedit
$ nix build .#pythonPackages.rchitect
$ nix build .#python3Packages.rchitect
$ nix build .#pythonPackages.radian
$ nix build .#python3Packages.radian
```

This is only my third `nixpkgs` PR, and my first PR touching Python or R. Please feel free to provide any feedback, even if you think it is just a nit! I'm also not sure if this PR is too large; should I make a PR just providing the Python packages, and then make a PR adding the wrapper if there is more discussion needed? Even if I just merged the Python packages, that would simplify my NixOS setup :smile: 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jdreaver committed:
radian: init at 0.5.11 and provide radianWrapper

Also:
lineedit: init at 0.1.6
rchitect: init at 0.3.32

jdreaver commented on pkgs/development/r-modules/wrapper-radian.nix:
```diff
@@ -0,0 +1,15 @@
+{ lib, runCommand, R, recommendedPackages, packages, makeWrapper, radian }:
```
This file was adapted from the adjacent `wrapper-rstudio.nix`. It is a lot simpler since we don't need to deal with Qt.

jdreaver committed:
Add myself as a maintainer

I have another open PR doing this, and I forgot that wasn't merged.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 137555` at 5988c7d2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>python38Packages.lineedit</li>
    <li>python38Packages.radian</li>
    <li>python38Packages.rchitect</li>
    <li>python39Packages.lineedit</li>
    <li>python39Packages.radian</li>
    <li>python39Packages.rchitect</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/radian/default.nix:35:0:
```
   |
35 |     description = "A 21 century R console.";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/rchitect/default.nix:35:0:
```
   |
35 |     description = "Interoperate R with Python";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/lineedit/default.nix:28:0:
```
   |
28 |     description = "A readline library based on prompt_toolkit which supports multiple modes.";
   | ^
```
</li>
  </ul>
</details>

jdreaver committed:
Add pythonImportsCheck to lineedit, rchitect, and radian

SuperSandro2000 commented on pkgs/development/r-modules/wrapper-radian.nix:
```diff
@@ -0,0 +1,15 @@
+  preferLocalBuild = true;
+  allowSubstitutes = false;
+
+  nativeBuildInputs = [makeWrapper];
+
+  buildInputs = [R] ++ recommendedPackages ++ packages;
```
```suggestion
  nativeBuildInputs = [ makeWrapper ];

  buildInputs = [ R ] ++ recommendedPackages ++ packages;
```

SuperSandro2000 commented on pkgs/development/python-modules/rchitect/default.nix:
```diff
@@ -0,0 +1,42 @@
+  propagatedBuildInputs = [
+    cffi
+    six
+  ] ++ lib.optionals isPy27 [
+    backports-shutil-which
```
We no longer support python27 for new packages.

SuperSandro2000 commented on pkgs/development/python-modules/rchitect/default.nix:
```diff
@@ -0,0 +1,42 @@
+    pytest-runner
+  ];
+
+  # Tests not included in PyPI release
```
Can we fetch the source instead?

SuperSandro2000 commented on pkgs/development/python-modules/rchitect/default.nix:
```diff
@@ -0,0 +1,42 @@
+    sha256 = "ghA7Go1E3TcGqgvY6CVP5UKN1S1GLwy5hmxYPrcoasM=";
+  };
+
+  buildInputs = [
+    pytest-runner
+  ];
+
```
```suggestion
```
please patch out the requirement on pytest-runner.

SuperSandro2000 commented on pkgs/development/python-modules/radian/default.nix:
```diff
@@ -0,0 +1,42 @@
+    sha256 = "hM2IDcBjOFIRC3SdNO9PSM8TqI1ga1GvCFQkqf4Tyto=";
+  };
+
+  buildInputs = [
+    pytest-runner
+  ];
+
```
```suggestion
```

SuperSandro2000 reviewed (changes requested):
Please try to enable tests for all packages.

jdreaver committed:
Don't support Python 2.7 for new package (rchitect)

jdreaver committed:
Attempt to get lineedit, rchitect, radian tests running

jdreaver committed:
Format whitespace in arrays

jdreaver commented:
Hey @SuperSandro2000, I addressed your comments, but I couldn't get the tests passing in these packages unfortunately. I changed the `doCheck` comments explaining what went wrong in each package, and I at least got the tests to the point where they could run by providing the correct `checkInputs`. The tests for these packages are quite impure, and require a terminal and a working R setup in the test environment. I suspect it would be quite a bit of work to get them running.

jonringer commented on pkgs/development/python-modules/radian/default.nix:
```diff
@@ -0,0 +1,61 @@
+  pythonImportsCheck = [ "radian" ];
+
+  meta = with lib; {
+    description = "A 21 century R console.";
```
```suggestion
    description = "A 21 century R console";
```

jonringer replied:
does this also need R at runtime? if so, you will likely need to patch the source so that it references a valid path, just having R available at build time won't mean it will be available at run time.

jonringer commented:
> Hey @SuperSandro2000, I addressed your comments, but I couldn't get the tests passing in these packages unfortunately. I changed the `doCheck` comments explaining what went wrong in each package, and I at least got the tests to the point where they could run by providing the correct `checkInputs`. The tests for these packages are quite impure, and require a terminal and a working R setup in the test environment. I suspect it would be quite a bit of work to get them running.

It is desirable to get them running. However, for nixpkgs, we are more or less just asserting that the package is working as upstream expected, not necessarily trying to assert that a release is flawless. Running a subset of tests I would also say is fine.

Another possibility (which would be a bit more work) would be to do a scenario using radian, and add that simple scenario using radian through `passthru.tests`, this would be a validation that it "works for nixpkgs" :)

SuperSandro2000 commented on pkgs/development/python-modules/lineedit/default.nix:
```diff
@@ -10,12 +10,16 @@ buildPythonPackage rec {
+  src = fetchFromGitHub {
+    owner = "randy3k";
+    repo = pname;
+    rev = "4262e9852aa50b09ff3f0569e1840b003ed00584";
```
```suggestion
    rev = "v${version}";
```

SuperSandro2000 commented on pkgs/development/python-modules/radian/default.nix:
```diff
@@ -12,16 +17,30 @@ buildPythonPackage rec {
+
+  checkInputs = [
+    pytestCheckHook
+    coverage
```
```suggestion
```
please patch out that this is required.

SuperSandro2000 commented on pkgs/development/python-modules/radian/default.nix:
```diff
@@ -12,16 +17,30 @@ buildPythonPackage rec {
+  src = fetchFromGitHub {
+    owner = "randy3k";
+    repo = pname;
+    rev = "c1579e6b3fa1c302961d08debe17232127616b9b";
```
```suggestion
    rev = "v${version}";
```

SuperSandro2000 commented on pkgs/development/python-modules/rchitect/default.nix:
```diff
@@ -1,34 +1,37 @@
+  src = fetchFromGitHub {
+    owner = "randy3k";
+    repo = pname;
+    rev = "0f3d5a191cca895a1cb9c41f8391520a08e8b9fc";
```
```suggestion
    rev = "v${version}";
```

natsukium commented:
closed by #217517
